overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -148,3 +151,62 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 		return fmt.Errorf("%s command is taking too long", inst.Action)
 	}
 }
+
+// NoAttributeError indicates that an interface attribute is not set.
+type NoAttributeError struct {
+	Attribute string
+}
+
+func (e *NoAttributeError) Error() string {
+	return fmt.Sprintf("unknown attribute %q", e.Attribute)
+}
+
+// isNoAttribute returns whether the provided error is a *NoAttributeError.
+func isNoAttribute(err error) bool {
+	_, ok := err.(*NoAttributeError)
+	return ok
+}
+
+func jsonRaw(v interface{}) *json.RawMessage {
+	data, err := json.Marshal(v)
+	if err != nil {
+		panic(fmt.Errorf("internal error: cannot marshal attributes: %v", err))
+	}
+	raw := json.RawMessage(data)
+	return &raw
+}
+
+// getAttribute unmarshals into result the value of the provided key from attributes map.
+// If the key does not exist, an error of type *NoAttributeError is returned.
+// The provided key may be formed as a dotted key path through nested maps.
+// For example, the "a.b.c" key describes the {a: {b: {c: value}}} map.
+func getAttribute(snapName string, subkeys []string, pos int, attrs map[string]interface{}, result interface{}) error {
+	value, ok := attrs[subkeys[pos]]
+	if !ok {
+		return &NoAttributeError{Attribute: strings.Join(subkeys[:pos+1], ".")}
+	}
+
+	if pos+1 == len(subkeys) {
+		raw, ok := value.(*json.RawMessage)
+		if !ok {
+			raw = jsonRaw(value)
+		}
+		if err := jsonutil.DecodeWithNumber(bytes.NewReader(*raw), &result); err != nil {
+			key := strings.Join(subkeys, ".")
+			return fmt.Errorf("internal error: cannot unmarshal snap %s attribute %q into %T: %s, json: %s", snapName, key, result, err, *raw)
+		}
+		return nil
+	}
+
+	attrsm, ok := value.(map[string]interface{})
+	if !ok {
+		raw, ok := value.(*json.RawMessage)
```

Comment:
all this clause doesn't seem covered

Reply:
Should be covered now with an extra test for `get` on interface attributes that's not a map.